In [39]:
%matplotlib inline
import numpy as np 
import pandas as pd
import sklearn as sk 
import matplotlib.pyplot as plt

## Pandas Dataframes

We will be using Pandas dataframes for much of this class to organize and sort through economic data. Pandas is one of the most widely used Python libraries in data science. It is mainly used for data cleaning, and with good reason: it’s very powerful and flexible, among many other things. 

### Creating dataframes

The rows and columns of a pandas dataframe are essentially a collection of lists stacked on top/next to each other. For example, if I wanted to store the top 10 movies and their ratings in a datatable, I could create 10 lists that each contain a rating and a corresponding title, and these lists would be the rows of the table:

In [53]:
top_10_movies = pd.DataFrame(data=np.array(
            [[9.2, 'The Shawshank Redemption (1994)'],
            [9.2, 'The Godfather (1972)'],
            [9., 'The Godfather: Part II (1974)'],
            [8.9, 'Pulp Fiction (1994)'],
            [8.9, "Schindler's List (1993)"],
            [8.9, 'The Lord of the Rings: The Return of the King (2003)'],
            [8.9, '12 Angry Men (1957)'],
            [8.9, 'The Dark Knight (2008)'],
            [8.9, 'Il buono, il brutto, il cattivo (1966)'],
            [8.8, 'The Lord of the Rings: The Fellowship of the Ring (2001)']]), columns=["Rating", "Movie"])
top_10_movies

Rating                                              Movie
0    9.2                    The Shawshank Redemption (1994)
1    9.2                               The Godfather (1972)
2    9.0                      The Godfather: Part II (1974)
3    8.9                                Pulp Fiction (1994)
4    8.9                            Schindler's List (1993)
5    8.9  The Lord of the Rings: The Return of the King ...
6    8.9                                12 Angry Men (1957)
7    8.9                             The Dark Knight (2008)
8    8.9             Il buono, il brutto, il cattivo (1966)
9    8.8  The Lord of the Rings: The Fellowship of the R...

Alternatively, we can store data in a dictionary instead of in lists. A dictionary keeps a mapping of keys to a set of values, and each key is unique. Using our top 10 movies example, we could create a dictionary that contains ratings a key, and movie titles as another key.

In [54]:
top_10_movies_dict = {"Rating" : [9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8], 
                      "Movie" : ['The Shawshank Redemption (1994)',
                                'The Godfather (1972)',
                                'The Godfather: Part II (1974)',
                                'Pulp Fiction (1994)',
                                "Schindler's List (1993)",
                                'The Lord of the Rings: The Return of the King (2003)',
                                '12 Angry Men (1957)',
                                'The Dark Knight (2008)',
                                'Il buono, il brutto, il cattivo (1966)',
                                'The Lord of the Rings: The Fellowship of the Ring (2001)']}

Now, we can use this dictionary to create a table with columns `Rating` and `Movie`

In [57]:
top_10_movies_2 = pd.DataFrame(data=top_10_movies_dict, columns=["Rating", "Movie"])
top_10_movies_2

Rating                                              Movie
0     9.2                    The Shawshank Redemption (1994)
1     9.2                               The Godfather (1972)
2     9.0                      The Godfather: Part II (1974)
3     8.9                                Pulp Fiction (1994)
4     8.9                            Schindler's List (1993)
5     8.9  The Lord of the Rings: The Return of the King ...
6     8.9                                12 Angry Men (1957)
7     8.9                             The Dark Knight (2008)
8     8.9             Il buono, il brutto, il cattivo (1966)
9     8.8  The Lord of the Rings: The Fellowship of the R...

Notice how both ways return the same table! However, the list method created the table by essentially taking the lists and making up the rows of the table, while the dictionary method took the keys from the dictionary to make up the columns of the table. In this way, dataframes can be viewed as a collection of basic data structures, either through collecting rows or columns. 

### Reading in Dataframes

Luckily for you, most datatables in this course will be premade and given to you in a form that is easily read into a pandas method, which creates the table for you. A common file type that is used for economic data is a Comma-Separated Values(.csv) file, which stores tabular data. It is not necessary for you to know exactly how .csv files store data, but you should know how to read a file in as a pandas dataframe. 

We will read in a .csv file that contains quarterly real GDI, real GDP, and nominal GDP data in the U.S. from 1947 to the present.

In [103]:
### Run this cell to read in the table
accounts = pd.read_csv("data/Quarterly_Accounts.csv")

The `pd.read_csv` function expects a path to a .csv file as its input, and will return a datatable created from the data contained in the csv.
We have provided `Quarterly_Accouunts.csv` in the data directory, which is all contained in the current working directory (aka the folder this assignment is contained in). For this reason, we must specify to the `read_csv` function that it should look for the csv in the data directory, and the `/` indicates that `Quarterly_Accounts.csv` can be found there. 

Here is a sample of some of the rows in this datatable:

In [104]:
accounts.head()

Year Quarter  Real GDI  Real GDP  Nominal GDP
0  1947      Q1    1912.5    1934.5        243.1
1  1947      Q2    1910.9    1932.3        246.3
2  1947      Q3    1914.0    1930.3        250.1
3  1947      Q4    1932.0    1960.7        260.3
4  1948      Q1    1984.4    1989.5        266.2

### Indexing Dataframes

Oftentimes, tables will contain a lot of extraneous data that muddles our datatables, making it more difficult to quickly and accurately obtain the data we need. To correct for this, we can select out columns or rows that we need by indexing our dataframes. 

The easiest way to index into a table is with square bracket notation. Suppose you wanted to obtain all of the Real GDP data from the data. Using a single pair of square brackets, you could index the table for `"Real GDP"`

In [72]:
## Run this cell and see what it outputs
accounts["Real GDP"]

0       1934.5
1       1932.3
2       1930.3
3       1960.7
4       1989.5
5       2021.9
6       2033.2
7       2035.3
8       2007.5
9       2000.8
10      2022.8
11      2004.7
12      2084.6
13      2147.6
14      2230.4
15      2273.4
16      2304.5
17      2344.5
18      2392.8
19      2398.1
20      2423.5
21      2428.5
22      2446.1
23      2526.4
24      2573.4
25      2593.5
26      2578.9
27      2539.8
28      2528.0
29      2530.7
        ...   
251    14541.9
252    14604.8
253    14745.9
254    14845.5
255    14939.0
256    14881.3
257    14989.6
258    15021.1
259    15190.3
260    15291.0
261    15362.4
262    15380.8
263    15384.3
264    15491.9
265    15521.6
266    15641.3
267    15793.9
268    15757.6
269    15935.8
270    16139.5
271    16220.2
272    16350.0
273    16460.9
274    16527.6
275    16547.6
276    16571.6
277    16663.5
278    16778.1
279    16851.4
280    16903.2
Name: Real GDP, dtype: float64

Notice how the above cell returns an array of all the real GDP values in their original order.
Now, if you wanted to get the first real GDP value from this array, you could index it with another pair of square brackets:

In [79]:
accounts["Real GDP"][0]

1934.5

Keep in mind that pandas dataframes, as well as many other data structures, are zero-indexed, meaning indexes start at 0 and end at the number of elements minus one. 

If you wanted to create a new datatable with select columns from the original table, you can index with double brackets.

In [88]:
## Note: .head() returns the first five rows of the table
accounts[["Year", "Quarter", "Real GDP", "Real GDI"]].head()

Year Quarter  Real GDP  Real GDI
0  1947      Q1    1934.5    1912.5
1  1947      Q2    1932.3    1910.9
2  1947      Q3    1930.3    1914.0
3  1947      Q4    1960.7    1932.0
4  1948      Q1    1989.5    1984.4

You can also use column indices instead of names.

In [89]:
accounts[[0, 1, 2, 3]].head()

Year Quarter  Real GDI  Real GDP
0  1947      Q1    1912.5    1934.5
1  1947      Q2    1910.9    1932.3
2  1947      Q3    1914.0    1930.3
3  1947      Q4    1932.0    1960.7
4  1948      Q1    1984.4    1989.5

Alternatively, you can also get rid of columns you dont need using `.drop()`

In [95]:
accounts.drop("Nominal GDP", axis=1).head()

Year Quarter  Real GDI  Real GDP
0  1947      Q1    1912.5    1934.5
1  1947      Q2    1910.9    1932.3
2  1947      Q3    1914.0    1930.3
3  1947      Q4    1932.0    1960.7
4  1948      Q1    1984.4    1989.5

Finally, you can use square bracket notation to index rows by their indices with a single set of brackets. You must specify a range of values for which you want to index. For example, if I wanted the 20th to 30th rows of `accounts`:

In [106]:
accounts[20:31]

Year Quarter  Real GDI  Real GDP  Nominal GDP
20  1952      Q1    2398.3    2423.5        360.2
21  1952      Q2    2412.6    2428.5        361.4
22  1952      Q3    2435.0    2446.1        368.1
23  1952      Q4    2509.5    2526.4        381.2
24  1953      Q1    2554.3    2573.4        388.5
25  1953      Q2    2572.2    2593.5        392.3
26  1953      Q3    2555.7    2578.9        391.7
27  1953      Q4    2504.1    2539.8        386.5
28  1954      Q1    2510.1    2528.0        385.9
29  1954      Q2    2514.5    2530.7        386.7
30  1954      Q3    2537.1    2559.4        391.6

### Filtering Data

As you can tell from the previous, indexing rows based on indices is only useful when you know the specific set of rows that you need, and you can only really get a range of entries. Working with data often involves huge datasets, making it inefficient and sometimes impossible to know exactly what indices to be looking at. On top of that, most data analysis concerns itself with looking for patterns or specific conditions in the data, which is impossible to look for with simple index based sorting.   

Thankfully, you can also use square bracket notation to filter out data based on a condition. Suppose we only wanted real GDP and nominal GDP data from the 21st century:

In [113]:
accounts[accounts["Year"] >= 2000][["Real GDP", "Nominal GDP"]]

Real GDP  Nominal GDP
212   12359.1      10031.0
213   12592.5      10278.3
214   12607.7      10357.4
215   12679.3      10472.3
216   12643.3      10508.1
217   12710.3      10638.4
218   12670.1      10639.5
219   12705.3      10701.3
220   12822.3      10834.4
221   12893.0      10934.8
222   12955.8      11037.1
223   12964.0      11103.8
224   13031.2      11230.1
225   13152.1      11370.7
226   13372.4      11625.1
227   13528.7      11816.8
228   13606.5      11988.4
229   13706.2      12181.4
230   13830.8      12367.7
231   13950.4      12562.2
232   14099.1      12813.7
233   14172.7      12974.1
234   14291.8      13205.4
235   14373.4      13381.6
236   14546.1      13648.9
237   14589.6      13799.8
238   14602.6      13908.5
239   14716.9      14066.4
240   14726.0      14233.2
241   14838.7      14422.3
..        ...          ...
251   14541.9      14566.5
252   14604.8      14681.1
253   14745.9      14888.6
254   14845.5      15057.7
255   14939.0      15230.2
256   14881.3      15238.4
257   14989.6      15460.9
258   15021.1      15587.1
259   15190.3      15785.3
260   15291.0      15973.9
261   15362.4      16121.9
262   15380.8      16227.9
263   15384.3      16297.3
264   15491.9      16475.4
265   15521.6      16541.4
266   15641.3      16749.3
267   15793.9      16999.9
268   15757.6      17031.3
269   15935.8      17320.9
270   16139.5      17622.3
271   16220.2      17735.9
272   16350.0      17874.7
273   16460.9      18093.2
274   16527.6      18227.7
275   16547.6      18287.2
276   16571.6      18325.2
277   16663.5      18538.0
278   16778.1      18729.1
279   16851.4      18905.5
280   16903.2      19057.7

[69 rows x 2 columns]

The `accounts` table is being indexed by the condition `accounts["Year"] >= 2000`, which returns a table where only rows that have a "Year" greater than $2000$ is returned. We then index this table with the double bracket notation from the previous section to only get the real GDP and nominal GDP columns.

Suppose now we wanted a table with data from the first quarter, and where the real GDP was less than 5000 or nominal GDP is greater than 15,000.

In [127]:
accounts[(accounts["Quarter"] == "Q1") & ((accounts["Real GDP"] < 5000) | (accounts["Nominal GDP"] > 15000))]

Year Quarter  Real GDI  Real GDP  Nominal GDP
0    1947      Q1    1912.5    1934.5        243.1
4    1948      Q1    1984.4    1989.5        266.2
8    1949      Q1    2001.5    2007.5        275.4
12   1950      Q1    2060.1    2084.6        281.2
16   1951      Q1    2281.0    2304.5        336.4
20   1952      Q1    2398.3    2423.5        360.2
24   1953      Q1    2554.3    2573.4        388.5
28   1954      Q1    2510.1    2528.0        385.9
32   1955      Q1    2661.6    2683.8        413.8
36   1956      Q1    2775.4    2770.0        440.5
40   1957      Q1    2862.0    2854.5        470.6
44   1958      Q1    2779.9    2772.7        468.4
48   1959      Q1    2976.5    2976.6        511.1
52   1960      Q1    3121.9    3123.2        543.3
56   1961      Q1    3109.9    3102.3        545.9
60   1962      Q1    3328.6    3336.8        595.2
64   1963      Q1    3469.1    3456.1        622.7
68   1964      Q1    3658.6    3672.7        671.1
72   1965      Q1    3885.5    3873.5        719.2
76   1966      Q1    4167.8    4201.9        797.3
80   1967      Q1    4286.5    4324.9        846.0
84   1968      Q1    4465.6    4490.6        911.1
88   1969      Q1    4665.4    4691.6        995.4
92   1970      Q1    4690.4    4707.1       1053.5
96   1971      Q1    4778.0    4834.3       1137.8
254  2010      Q3   14839.5   14845.5      15057.7
255  2010      Q4   14904.9   14939.0      15230.2
256  2011      Q1   14924.4   14881.3      15238.4
257  2011      Q2   14996.1   14989.6      15460.9
258  2011      Q3   15093.1   15021.1      15587.1
259  2011      Q4   15217.0   15190.3      15785.3
260  2012      Q1   15500.4   15291.0      15973.9
261  2012      Q2   15522.8   15362.4      16121.9
262  2012      Q3   15517.1   15380.8      16227.9
263  2012      Q4   15650.6   15384.3      16297.3
264  2013      Q1   15642.7   15491.9      16475.4
265  2013      Q2   15719.8   15521.6      16541.4
266  2013      Q3   15752.0   15641.3      16749.3
267  2013      Q4   15851.3   15793.9      16999.9
268  2014      Q1   15912.8   15757.6      17031.3
269  2014      Q2   16136.1   15935.8      17320.9
270  2014      Q3   16327.9   16139.5      17622.3
271  2014      Q4   16520.8   16220.2      17735.9
272  2015      Q1   16599.6   16350.0      17874.7
273  2015      Q2   16700.6   16460.9      18093.2
274  2015      Q3   16726.7   16527.6      18227.7
275  2015      Q4   16789.8   16547.6      18287.2
276  2016      Q1   16776.1   16571.6      18325.2
277  2016      Q2   16783.0   16663.5      18538.0
278  2016      Q3   16953.0   16778.1      18729.1
279  2016      Q4   16882.1   16851.4      18905.5
280  2017      Q1   16992.1   16903.2      19057.7

Many different conditions can be included to filter, and you can use `&` and `|` operators to connect them together. Make sure to include parantheses for each condition!

Another way to reorganize data to make it more convenient is to sort the data by the values in a specific column. For example, if we wanted to find the highest real GDP since 1947, we could sort the table for real GDP:

In [134]:
accounts.sort_values("Real GDP")

Year Quarter  Real GDI  Real GDP  Nominal GDP
2    1947      Q3    1914.0    1930.3        250.1
1    1947      Q2    1910.9    1932.3        246.3
0    1947      Q1    1912.5    1934.5        243.1
3    1947      Q4    1932.0    1960.7        260.3
4    1948      Q1    1984.4    1989.5        266.2
9    1949      Q2    1995.9    2000.8        271.7
11   1949      Q4    1979.6    2004.7        271.0
8    1949      Q1    2001.5    2007.5        275.4
5    1948      Q2    2030.2    2021.9        272.9
10   1949      Q3    2007.9    2022.8        273.3
6    1948      Q3    2031.5    2033.2        279.5
7    1948      Q4    2041.6    2035.3        280.7
12   1950      Q1    2060.1    2084.6        281.2
13   1950      Q2    2144.4    2147.6        290.7
14   1950      Q3    2225.9    2230.4        308.5
15   1950      Q4    2268.9    2273.4        320.3
16   1951      Q1    2281.0    2304.5        336.4
17   1951      Q2    2321.3    2344.5        344.5
18   1951      Q3    2362.0    2392.8        351.8
19   1951      Q4    2382.7    2398.1        356.6
20   1952      Q1    2398.3    2423.5        360.2
21   1952      Q2    2412.6    2428.5        361.4
22   1952      Q3    2435.0    2446.1        368.1
23   1952      Q4    2509.5    2526.4        381.2
28   1954      Q1    2510.1    2528.0        385.9
29   1954      Q2    2514.5    2530.7        386.7
27   1953      Q4    2504.1    2539.8        386.5
30   1954      Q3    2537.1    2559.4        391.6
24   1953      Q1    2554.3    2573.4        388.5
26   1953      Q3    2555.7    2578.9        391.7
..    ...     ...       ...       ...          ...
244  2008      Q1   14842.2   14889.5      14668.4
246  2008      Q3   14767.0   14891.6      14843.0
242  2007      Q3   14822.4   14938.5      14569.7
255  2010      Q4   14904.9   14939.0      15230.2
245  2008      Q2   14832.4   14963.4      14813.0
257  2011      Q2   14996.1   14989.6      15460.9
243  2007      Q4   14816.6   14991.8      14685.3
258  2011      Q3   15093.1   15021.1      15587.1
259  2011      Q4   15217.0   15190.3      15785.3
260  2012      Q1   15500.4   15291.0      15973.9
261  2012      Q2   15522.8   15362.4      16121.9
262  2012      Q3   15517.1   15380.8      16227.9
263  2012      Q4   15650.6   15384.3      16297.3
264  2013      Q1   15642.7   15491.9      16475.4
265  2013      Q2   15719.8   15521.6      16541.4
266  2013      Q3   15752.0   15641.3      16749.3
268  2014      Q1   15912.8   15757.6      17031.3
267  2013      Q4   15851.3   15793.9      16999.9
269  2014      Q2   16136.1   15935.8      17320.9
270  2014      Q3   16327.9   16139.5      17622.3
271  2014      Q4   16520.8   16220.2      17735.9
272  2015      Q1   16599.6   16350.0      17874.7
273  2015      Q2   16700.6   16460.9      18093.2
274  2015      Q3   16726.7   16527.6      18227.7
275  2015      Q4   16789.8   16547.6      18287.2
276  2016      Q1   16776.1   16571.6      18325.2
277  2016      Q2   16783.0   16663.5      18538.0
278  2016      Q3   16953.0   16778.1      18729.1
279  2016      Q4   16882.1   16851.4      18905.5
280  2017      Q1   16992.1   16903.2      19057.7

[281 rows x 5 columns]

But wait! The table looks like it's sorted in increasing order. This is because `sort_values` defaults to ordering the column in ascending order. To correct this, add in the extra optional parameter

In [135]:
accounts.sort_values("Real GDP", ascending=False)

Year Quarter  Real GDI  Real GDP  Nominal GDP
280  2017      Q1   16992.1   16903.2      19057.7
279  2016      Q4   16882.1   16851.4      18905.5
278  2016      Q3   16953.0   16778.1      18729.1
277  2016      Q2   16783.0   16663.5      18538.0
276  2016      Q1   16776.1   16571.6      18325.2
275  2015      Q4   16789.8   16547.6      18287.2
274  2015      Q3   16726.7   16527.6      18227.7
273  2015      Q2   16700.6   16460.9      18093.2
272  2015      Q1   16599.6   16350.0      17874.7
271  2014      Q4   16520.8   16220.2      17735.9
270  2014      Q3   16327.9   16139.5      17622.3
269  2014      Q2   16136.1   15935.8      17320.9
267  2013      Q4   15851.3   15793.9      16999.9
268  2014      Q1   15912.8   15757.6      17031.3
266  2013      Q3   15752.0   15641.3      16749.3
265  2013      Q2   15719.8   15521.6      16541.4
264  2013      Q1   15642.7   15491.9      16475.4
263  2012      Q4   15650.6   15384.3      16297.3
262  2012      Q3   15517.1   15380.8      16227.9
261  2012      Q2   15522.8   15362.4      16121.9
260  2012      Q1   15500.4   15291.0      15973.9
259  2011      Q4   15217.0   15190.3      15785.3
258  2011      Q3   15093.1   15021.1      15587.1
243  2007      Q4   14816.6   14991.8      14685.3
257  2011      Q2   14996.1   14989.6      15460.9
245  2008      Q2   14832.4   14963.4      14813.0
255  2010      Q4   14904.9   14939.0      15230.2
242  2007      Q3   14822.4   14938.5      14569.7
246  2008      Q3   14767.0   14891.6      14843.0
244  2008      Q1   14842.2   14889.5      14668.4
..    ...     ...       ...       ...          ...
26   1953      Q3    2555.7    2578.9        391.7
24   1953      Q1    2554.3    2573.4        388.5
30   1954      Q3    2537.1    2559.4        391.6
27   1953      Q4    2504.1    2539.8        386.5
29   1954      Q2    2514.5    2530.7        386.7
28   1954      Q1    2510.1    2528.0        385.9
23   1952      Q4    2509.5    2526.4        381.2
22   1952      Q3    2435.0    2446.1        368.1
21   1952      Q2    2412.6    2428.5        361.4
20   1952      Q1    2398.3    2423.5        360.2
19   1951      Q4    2382.7    2398.1        356.6
18   1951      Q3    2362.0    2392.8        351.8
17   1951      Q2    2321.3    2344.5        344.5
16   1951      Q1    2281.0    2304.5        336.4
15   1950      Q4    2268.9    2273.4        320.3
14   1950      Q3    2225.9    2230.4        308.5
13   1950      Q2    2144.4    2147.6        290.7
12   1950      Q1    2060.1    2084.6        281.2
7    1948      Q4    2041.6    2035.3        280.7
6    1948      Q3    2031.5    2033.2        279.5
10   1949      Q3    2007.9    2022.8        273.3
5    1948      Q2    2030.2    2021.9        272.9
8    1949      Q1    2001.5    2007.5        275.4
11   1949      Q4    1979.6    2004.7        271.0
9    1949      Q2    1995.9    2000.8        271.7
4    1948      Q1    1984.4    1989.5        266.2
3    1947      Q4    1932.0    1960.7        260.3
0    1947      Q1    1912.5    1934.5        243.1
1    1947      Q2    1910.9    1932.3        246.3
2    1947      Q3    1914.0    1930.3        250.1

[281 rows x 5 columns]

Now we can clearly see that the highest real GDP was attained in the first quarter of this year, and had a value of 16903.2

### Useful Functions for Numeric Data

Here are a few useful functions when dealing with numeric data columns.
To find the minimum value in a column, call `min()` on a column of the table.

In [142]:
accounts["Real GDP"].min()

1930.3

To find the maximum value, call `max()`.

In [143]:
accounts["Nominal GDP"].max()

19057.700000000001

And to find the average value of a column, use `mean()`.

In [144]:
accounts["Real GDI"].mean()

7890.370462633456